ocaml/xapi/stream_vdi.ml

```diff
@@ -100,6 +103,105 @@ let write_block ~__context filename buffer ofd len =
+  if Astring.String.is_prefix ~affix:nbd_device_prefix path && is_nbd_device path then begin
+    let nbd_number =
+      Astring.String.with_range ~first:(String.length nbd_device_prefix) path
+    in
+    let { path; exportname } =
+      (* persistent_nbd_info_dir is written from nbd_client_manager.py as part of VBD plug*)
+      let persistent_nbd_info_dir = "/var/run/nonpersistent/nbd" in
+      let filename = persistent_nbd_info_dir ^ "/" ^ nbd_number in
+      Xapi_stdext_unix.Unixext.string_of_file filename
+      |> Jsonrpc.of_string
+      |> nbd_connect_info_of_rpc
+    in
+    Some (path, exportname)
+  end else None
+
+type extent = {
+  flags : int32;
+  length : int64;
+} [@@deriving rpc]
+
+type extent_list = extent list [@@deriving rpc]
+
+(* Flags for extents returned for the base:allocation NBD metadata context,
+   documented at https://github.com/NetworkBlockDevice/nbd/blob/master/doc/proto.md *)
+let flag_hole = 1l
+let flag_zero = 2l
+
+(** [descriptor_list] should be a list of non-overlapping extents, ordered from
+    lowest offset to highest.
+    [offset] is the current offset needed to translate from relative extents to 
+    absolute chunk numbers *)
+let get_chunk_numbers_in_increasing_order descriptor_list offset =
+  (* Output increasing range includes start and end points *)
+  let rec range acc start_chunk end_chunk =
+    if end_chunk < start_chunk then acc
+    else range (end_chunk::acc) start_chunk Int64.(add end_chunk minus_one)
+  in
+
+  let is_empty e =
+    let has_flag flag =
+      Int32.logand e.flags flag = flag
+    in
+    (* We assume the destination is prezeroed, so we do not have to copy zeroed extents *)
+    (has_flag flag_hole) || (has_flag flag_zero)
+  in
+
+  (* Output increasing chunks numbers covered by this descriptor if it is not empty *)
+  let get_non_empty_chunks descriptor offset increment =
+    match (is_empty descriptor) with
+      | false ->
+        let start_chunk = Int64.div offset increment in
+        (* If the chunk ends at the boundary don't include the next chunk *)
+        let end_chunk = Int64.(div (add (add descriptor.length offset) minus_one) increment) in
+        let chunks = range [] start_chunk end_chunk in
+        chunks, descriptor.length
+      | true -> [], descriptor.length
+  in
+
+  (* Only compare to most recent chunk added *)
+  let add_new acc b =
+    match acc with
+    | hd :: _ when Int64.equal b hd -> acc
+    | _ -> b::acc
+  in
+
+  (* Output decreasing chunk numbers, x should be increasing and acc should be decreasing *)
+  let rec add_unique_chunks acc = function
+    | [] -> acc
+    | x::xs -> add_unique_chunks (add_new acc x) xs
+  in
+
+  (* This works in reverse order *)
+  let rec process acc offset = function
+    | [] -> acc
+    | x::xs ->
+      let chunks, add_offset = get_non_empty_chunks x offset chunk_size in
+      process (add_unique_chunks acc chunks) (Int64.add offset add_offset) xs
```

Comment:
I think this is equivalent to:
`process (List.fold_left add_new_unique acc chunks) (Int64.add offset add_offset)`.

I'm still thinking how else this could be simplified, it should be possible to implement this with fewer lines of code.

Reply:
Agreed - however I think it's not worth spending too much time making this PR perfect, because hopefully both branches of this code will be rewritten soon if/when we implement generic CBT for all SRs - in that case we'll use CBT instead of NBD and will have only one branch.

So if the above issues do not cause performance problems (e.g. quadratic complexity) or bugs in case of some corner cases, I think we should merge it?

Reply:
I couldn't find a situation where this code would give incorrect results so far, although would've been nice to have unit tests.
Does the quicktest construct a lot of scenarios with holes/zeroes and other corner cases?

Reply:
Yes. It test empty VDI, full random VDI, random VDI with random number of random extents (random offset, length, data).

Reply:
Stefano has run it 100 times and it did not fail :)